### PR TITLE
Add Facts::Appointments and Dimensions::State

### PR DESCRIPTION
**Pull request by dwhenry**

### CODE REVIEW

File: db/migrate/20160307111110_create_facts_appointments.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CreateFactsAppointments < ActiveRecord::Migration
+  def change
+    create_table :facts_appointments do |t|
+      t.references :dimensions_date, index: true, foreign_key: true
+      # t.references :dimensions_state, index: true, foreign_key: true
```
Comment (andrewgarner) [inline db/migrate/20160307111110_create_facts_appointments.rb]: We shouldn't be committing commented out code.

File: features/step_definitions/appointment_steps.rb.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  @count_by_state = Facts::Appointment.find_by_sql([sql, begin_date, end_date]).map do |appointment|
+    {
+      'appointment_status' => appointment.name,
+      'volume_of_appointments' => appointment.appointments.to_s
+    }
+  end
+end
+
+When(/^appointment data is extracted from the booking api$/) do
+  # build required states first - should this have it's own step??
```
Comment (andrewgarner) [inline features/step_definitions/appointment_steps.rb.rb]: Maybe these should be seeded instead?

File: spec/models/dimensions/state_spec.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe Dimensions::State, type: :model do
+  it { is_expected.to validate_presence_of(:name) }
+
+  it 'can only have a single default state' do
```
Comment (andrewgarner) [inline spec/models/dimensions/state_spec.rb]: I would extract the context for the test to something like this:

```ruby
subject(:state) { described_class.new(name: double, default: default) }

context 'when a default state already exists' do
  before { described_class.create(name: double, default: true) }

  let(:default) { true }

  specify 'only one default state can exist' do
    is_expected.to_not be_valid
  end
end
```

File: features/appointment/importing.feature [hunk 4]
```diff
@@ -0,0 +1,30 @@
+Feature: Appointment status
+  To enable reporting based on appointment outcome
+
+  @vcr_booking_bug_all
+  Scenario: Appointmonets are imported with status values
```
Comment (andrewgarner) [inline features/appointment/importing.feature]: /s/Appointmonets/Appointments

File: features/appointment/importing.feature [hunk 5]
```diff
@@ -0,0 +1,30 @@
+Feature: Appointment status
+  To enable reporting based on appointment outcome
```
Comment (andrewgarner) [inline features/appointment/importing.feature]: What has to happen?

File: features/appointment/importing.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  @vcr_booking_bug_all
+  Scenario: Appointmonets are imported with status values
+    Given the booking system contains booking data
+    And our stored dataset has date dimensions for the period "5/2/2015" to "20/3/2015"
+    When appointment data is extracted from the booking api
+    And I query appointments by status between "5/2/2015" and "20/3/2015"
```
Comment (andrewgarner) [inline features/appointment/importing.feature]: Who am "I"? We've given no context in the feature definition.

Comment (andrewgarner) [inline features/appointment/importing.feature]: What has querying got to do with an importing feature?

File: app/services/booking_bug/appointments.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+module BookingBug
+  class Appointments < Base
+    MAX_UPDATE_DAYS = 7 # used to limit number of entries queried to determine if an update ahs occurred.
+
+    # rubocop:disable MethodLength, AbcSize
```
Comment (andrewgarner) [inline app/services/booking_bug/appointments.rb]: I feel like we should be listening to Rubocop here. There is a lot of behaviour in this method.

Comment (dwhenry) [inline app/services/booking_bug/appointments.rb]: I think we can definitely come up with a better approach, potential one that can re-use parts between the different importers.

I will start thinking about this as the next item of work as don't think trying to refactor before merging makes any sense.

If you have any ideas/plans on how this should be done, give me a shout.

**Pull request merged**